tiledb/array.py

```diff
@@ -243,3 +251,748 @@ def index_domain_subarray(array, dom, idx: tuple):
+
+        :rtype: tuple(tuple(numpy scalar, numpy scalar), ...)
+        :return: A list of (inclusive) domain extent tuples, that contain all
+            nonempty cells
+
+        """
+        results = list()
+        dom = self.schema.domain
+
+        for dim_idx in range(dom.ndim):
+            dim_dtype = dom.dim(dim_idx).dtype
+
+            if self.array._non_empty_domain_is_empty(dim_idx, dim_dtype, self.ctx):
+                results.append((None, None))
+                continue
+
+            res_x, res_y = self.array._non_empty_domain(dim_idx, dim_dtype)
+
+            if dim_dtype == np.bytes_:
+                # convert to bytes if needed
+                res_x = bytes(res_x, encoding="utf8")
+                res_y = bytes(res_y, encoding="utf8")
+
+            if np.issubdtype(dim_dtype, np.datetime64):
+                # convert to np.datetime64
+                date_unit = np.datetime_data(dim_dtype)[0]
+                res_x = np.datetime64(res_x, date_unit)
+                res_y = np.datetime64(res_y, date_unit)
+
+            results.append((res_x, res_y))
+
+        # if result only contains (None, None) tuples just return None
+        for x in results:
+            if x != (None, None):
+                return tuple(results)
+
+        return None
+
+    def consolidate(self, config=None, key=None, fragment_uris=None, timestamp=None):
+        """
+        Consolidates fragments of an array object for increased read performance.
+
+        Overview: https://docs.tiledb.com/main/concepts/internal-mechanics/consolidation
+
+        :param tiledb.Config config: The TileDB Config with consolidation parameters set
+        :param key: (default None) encryption key to decrypt an encrypted array
+        :type key: str or bytes
+        :param fragment_uris: (default None) Consolidate the array using a list of fragment _names_ (note: the `__ts1_ts2_<label>_<ver>` fragment name form alone, not the full path(s))
+        :param timestamp: (default None) If not None, consolidate the array using the given tuple(int, int) UNIX seconds range (inclusive). This argument will be ignored if `fragment_uris` is passed.
+        :type timestamp: tuple (int, int)
+        :raises: :py:exc:`tiledb.TileDBError`
+
+        Rather than passing the timestamp into this function, it may be set with
+        the config parameters `"sm.vacuum.timestamp_start"`and
+        `"sm.vacuum.timestamp_end"` which takes in a time in UNIX seconds. If both
+        are set then this function's `timestamp` argument will be used.
+
+        """
+        if self.mode == "r":
+            raise tiledb.TileDBError(
+                "cannot consolidate array opened in readonly mode (mode='r')"
+            )
+
+        if config is None:
+            config = tiledb.Config(self.ctx.config())
+
+        if key is not None:
+            config["sm.encryption_type"] = "AES_256_GCM"
+            config["sm.encryption_key"] = key
```

Comment:
```suggestion
        if key is not None:
            config = tiledb.Config(config or self.ctx.config())
            config["sm.encryption_type"] = "AES_256_GCM"
            config["sm.encryption_key"] = key
```

This will avoid needlessly allocating a config, and modifying the user-provided config if encryption is used.